wifiphisher/common/victim.py

```diff
@@ -0,0 +1,171 @@
+
+        Victims_instance = Victims.getInstance()
+        if vmac_address in Victims_instance.victims_dic.keys():
+            Victims_instance.victims_dic[vmac_address].ip_address = ip_address
+        else:
+            raise Exception("Error: No such mac address exists in dictionary")
+
+
+class Victims():
+    """
+    Singleton class that manages all of the victims
+    """
+
+    # Instance will be stored here.
+    __instance = None
+
+    @staticmethod
+    def getInstance():
+        if Victims.__instance is None:
+            Victims()
+        return Victims.__instance
+
+    def __init__(self):
+        if Victims.__instance is not None:
+            raise Exception("Error: This class is a singleton!")
+        else:
+            Victims.__instance = self
+            self.victims_dic = {}
+
+    def add_to_victim_dic(self, Victim):
+        """
+        Add new victim to the dictionary
+        """
+
+        self.victims_dic[Victim.vmac_address] = Victim
+
+    def remove_from_victim_dic_bymac(self, vmac_address):
+        """
+        Delete victim by victim mac address
+        """
+
+        if vmac_address in self.victims_dic.keys():
+            self.victims_dic.pop(vmac_address, 0)
+        else:
+            raise Exception("Error: Victim to be deleted Not found")
+
+    def get_print_representation(self):
+        """
+        Returns the victim dictionary in str format in order to be printed
+
+        :param self: Victims instance
+        :type self: Victims
+        :rtype str
+        """
+
+        out = ""
+        for value in self.victims_dic.values():
+            out = out + value.vmac_address + '\t' + value.ip_address + '\t' \
+                     + value.os + '\t' + value.vendor + '\n'
+        return out
+
+    def associate_victim_ip_to_os(self, ip_address, url):
+        """
+        receives a victims ip address and request as input, finds the
+        corresponding os by reading the initial requests file and then accesses
+        the victim dictionary and changes the os for the victim with the
+        corresponding ip address
+
+        :param self: Victims instance
+        :type self: Victims
+        :param ip_address: ip address of the victim
+        :type ip_address: str
+        :param url: request of the victim
+        :type url: str
+        """
+
+        path_to_url_file = constants.URL_TO_OS_FILE
+        with open(path_to_url_file, "r") as url_file:
+            for line in url_file:
+                line = line.split("|")
+                url_check = line[1].strip()
+                os = line[0].strip()
+                if url_check in url:
+                    for key in self.victims_dic.keys():
+                        if ip_address == self.victims_dic[key].ip_address:
+                            self.victims_dic[key].os = os
+
+    def find_devices_by_os(self, os):
+        """
+        takes as input a specific os, prints devices found with that os and
+        returns their number
```

Comment:
where does this function get called?

Reply:
its not, its for the future if you want to print stats for the victims, i can remove it if you want

Reply:
Please remove it as it just adds noise to code. If it is needed we can always implement it in the future.

Reply:
done